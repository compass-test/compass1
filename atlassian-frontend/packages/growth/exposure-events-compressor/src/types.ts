export type Reason =
  | 'OFF'
  | 'FALLTHROUGH'
  | 'RULE_MATCH'
  | 'TARGET_MATCH'
  | 'INELIGIBLE'
  | 'SIMPLE_EVAL'
  | 'ERROR';

export type CriteriaFn = (event: ExposureEvent) => boolean;

export type ExposureEvent = {
  action: string;
  actionSubject: string;
  eventType: string;
  attributes: ExposureEventAttributes;
  tags: string[];
  source: string;
};

export type CompressedEvaluations = {
  action: string;
  actionSubject: string;
  eventType: string;
  attributes: {
    evaluations: EvaluationAttributes[];
  };
  tags: string[];
  source: string;
};

export type ExposureEventAttributes = {
  flagKey: string;
  value: boolean | string | object;
  reason?: Reason;
  ruleId?: string;
};

export type EvaluationAttributes = {
  key: string;
  value: boolean | string | object;
  reason?: Reason;
  ruleId?: string;
};
