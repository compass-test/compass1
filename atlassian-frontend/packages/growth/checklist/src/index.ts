import AccordionItem, {
  AccordionDescriptionWrapper,
  ACCORDION_CLASS_NAME,
} from './components/accordion-item/AccordionItem';
import AccordionItemButton from './components/accordion-item/AccordionItemButton';
import AccordionItemLink from './components/accordion-item/AccordionItemLink';
import {
  AccordionHeaderWrapper,
  AccordionItemButtonText,
  AccordionItemContainer,
} from './components/accordion-item/styled';
import ChecklistItem, { Task } from './components/checklist-item/ChecklistItem';
import ChecklistContainer, {
  ChecklistContainerProps,
} from './components/checklist-container/ChecklistContainer';
import ChecklistProgressBar, {
  ChecklistProgressBarProps,
} from './components/checklist-progressbar/ChecklistProgressBar';
import ChecklistButton, {
  ChecklistButtonProps,
} from './components/checklist-button/ChecklistButton';
import Checklist, { ChecklistProps } from './components/checklist/Checklist';
import ChecklistBackground from './components/checklist/ChecklistBackground';
import ChecklistFooter, {
  ChecklistFooterProps,
} from './components/checklist/ChecklistFooter';

export {
  AccordionItem,
  AccordionItemButton,
  AccordionItemButtonText,
  AccordionItemLink,
  AccordionItemContainer,
  AccordionDescriptionWrapper,
  AccordionHeaderWrapper,
  ACCORDION_CLASS_NAME,
  ChecklistItem,
  ChecklistContainer,
  ChecklistProgressBar,
  ChecklistButton,
  Checklist,
  ChecklistBackground,
  ChecklistFooter,
};
export type {
  Task,
  ChecklistContainerProps,
  ChecklistProgressBarProps,
  ChecklistButtonProps,
  ChecklistProps,
  ChecklistFooterProps,
};
