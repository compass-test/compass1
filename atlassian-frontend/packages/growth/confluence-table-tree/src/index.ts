export { TreeStates } from './types';
export type { AnalyticsWebClient } from './types';
export type { PageTreeItemProperties } from './ui/page-tree-for-embedded-pages/types';
export { default as ConfluencePageTree } from './components/ConfluencePageTree';
export { default } from './components/ConfluencePageTree';
export { AnalyticsProvider } from './common/AnalyticsProvider';
export { ConfluencePageTreeWithoutHeading } from './ui/confluence-page-tree';
export { default as DraftsToggle } from './ui/drafts-toggle';
export type { PageActionHandler } from './ui/confluence-page-tree/types';
export {
  ConfluencePageTreeProvider,
  useConfluencePageTree,
} from './controllers/page-tree';
export { pageAnalyticsAttributes } from './ui/page-tree-for-embedded-pages/utils';
