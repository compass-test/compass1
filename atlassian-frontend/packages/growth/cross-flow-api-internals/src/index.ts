export {
  BaseCrossFlowApiProvider,
  useCrossFlow,
  withCrossFlow,
} from './lib/context';
export { Journeys, Targets, Reasons } from './lib/types';
export type {
  ExperimentalOptions,
  TargetType,
  JourneyTargetProductDefinition,
  Options,
  CompletionStatus,
  OnOpen,
  ReasonType,
  CrossFlowContextType,
  WithCrossFlowProps,
} from './lib/types';
