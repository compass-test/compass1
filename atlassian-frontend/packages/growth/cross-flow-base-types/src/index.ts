export { CrossFlowExtensions } from './CrossFlowExtensions';
export type {
  CrossFlowExtensionsOptions,
  JSONValue,
  Values,
  DataPlugin,
  DataPluginNameType,
  DataPluginImplementationType,
  PluginCollection,
} from './types';
