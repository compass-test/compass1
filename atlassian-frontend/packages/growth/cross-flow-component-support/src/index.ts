export {
  useCrossFlow,
  withCrossFlow,
  Journeys,
  Targets,
  Reasons,
} from '@atlassiansox/cross-flow-api-internals';
export type {
  CompletionStatus,
  CrossFlowContextType,
  Options,
  ReasonType,
  WithCrossFlowProps,
  TargetType,
} from '@atlassiansox/cross-flow-api-internals';
