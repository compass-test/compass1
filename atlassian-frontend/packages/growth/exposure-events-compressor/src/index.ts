export { canCompress, buildCompressionFunction } from './main';
