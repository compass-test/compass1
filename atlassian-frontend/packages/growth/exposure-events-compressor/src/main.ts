import {
  CompressedEvaluations,
  EvaluationAttributes,
  ExposureEvent,
} from './types';

export const canCompress = (event: ExposureEvent): boolean =>
  event.actionSubject === 'feature' &&
  event.action === 'exposed' &&
  event.eventType === 'operational' &&
  event.attributes &&
  event.tags &&
  event.tags.includes('measurement') &&
  event.tags.includes('autoExposure');

export const compressor = (
  events: ExposureEvent[],
  batchSize: number = 50,
): CompressedEvaluations[] => {
  if (!events || events.length <= 0) {
    return [];
  }

  if (batchSize <= 0) {
    throw new Error('Batch size must be greater than 0');
  }
  // Extract evaluations and reduce into a single array
  const flagEvaluations: EvaluationAttributes[] = events.map((currentEvent) => {
    const { flagKey, value, reason, ruleId } = currentEvent.attributes;

    return {
      key: flagKey,
      value,
      ...(reason && { reason }),
      ...(ruleId && { ruleId }),
    };
  });

  // Batch evaluations into specified batchSize
  return createBatches(flagEvaluations, batchSize);
};

const createBatches = (
  array: EvaluationAttributes[],
  batchSize: number,
): CompressedEvaluations[] => {
  let compressedEvaluations = [];

  for (let i = 0; i < array.length; i += batchSize) {
    let evaluationBatch = array.slice(i, i + batchSize);
    compressedEvaluations.push({
      actionSubject: 'features',
      action: 'exposed',
      eventType: 'operational',
      attributes: { evaluations: evaluationBatch },
      tags: ['measurement', 'autoExposure'],
      source: '@atlassian/exposure-events-compressor',
    });
  }
  return compressedEvaluations;
};

export const buildCompressionFunction = (batchSize?: number) => (
  events: ExposureEvent[],
): CompressedEvaluations[] => compressor(events, batchSize);
