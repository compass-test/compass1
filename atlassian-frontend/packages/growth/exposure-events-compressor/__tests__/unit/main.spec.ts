import { CompressionRule } from '@atlassiansox/analytics-web-client';

import { buildCompressionFunction, canCompress } from '../../src';
import invalidEvents from '../fixtures/invalid-feature-exposed-events.json';
import validEvents from '../fixtures/valid-feature-exposed-events.json';

const getCompressorForBatchSize = (batchSize: number) =>
  new CompressionRule(canCompress, buildCompressionFunction(batchSize));

describe('compressor', () => {
  describe('events argument', () => {
    test('should compress all normal events', () => {
      expect(
        getCompressorForBatchSize(50).compress(validEvents),
      ).toMatchSnapshot();
    });
    test('should default to uncompressed events for invalid events and skip those with no attributes', () => {
      expect(
        getCompressorForBatchSize(50).compress(invalidEvents),
      ).toMatchSnapshot();
    });

    test('should return empty list if events list is empty', () => {
      expect(getCompressorForBatchSize(50).compress([])).toMatchSnapshot([]);
    });
  });

  describe('batchSize argument', () => {
    test('should compress all normal events into batches of 2 events', () => {
      expect(
        getCompressorForBatchSize(2).compress(validEvents),
      ).toMatchSnapshot();
    });

    test('should throw if batchSize is invalid', () => {
      expect(() => getCompressorForBatchSize(-1).compress(validEvents)).toThrow(
        'Batch size must be greater than 0',
      );
      expect(() => getCompressorForBatchSize(0).compress(validEvents)).toThrow(
        'Batch size must be greater than 0',
      );
    });

    test('should compress to one batch if batchSize is larger than events list length', () => {
      expect(
        getCompressorForBatchSize(10).compress(validEvents),
      ).toMatchSnapshot();
    });

    test('should compress to correct number of batches when events list length is not a multiple of batchSize', () => {
      expect(
        getCompressorForBatchSize(4).compress(validEvents),
      ).toMatchSnapshot();
    });
  });
});
