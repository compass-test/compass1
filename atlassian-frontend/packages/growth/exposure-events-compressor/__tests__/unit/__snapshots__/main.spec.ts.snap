// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compressor batchSize argument should compress all normal events into batches of 2 events 1`] = `
Array [
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "pin-2.0.refinementbar-dropdown",
          "reason": "SIMPLE_EVAL",
          "value": false,
        },
        Object {
          "key": "jsd.queues.keyboard-shortcuts",
          "reason": "ERROR",
          "ruleId": "111-bbbbb-ccc",
          "value": true,
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "empanada.nin-export-issues-dashboard-gadgets_yb8zi",
          "reason": "RULE_MATCH",
          "ruleId": "111-bcbcbcb-dddd",
          "value": false,
        },
        Object {
          "key": "nav-ux-enable-async-edit-filter-modal",
          "value": true,
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "project.config.pce.project-archival",
          "value": true,
        },
        Object {
          "key": "jsw.frontend.growth.experiments.persistent.upgrade",
          "reason": "FALLTHROUGH",
          "value": "not-enrolled",
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
]
`;

exports[`compressor batchSize argument should compress to correct number of batches when events list length is not a multiple of batchSize 1`] = `
Array [
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "pin-2.0.refinementbar-dropdown",
          "reason": "SIMPLE_EVAL",
          "value": false,
        },
        Object {
          "key": "jsd.queues.keyboard-shortcuts",
          "reason": "ERROR",
          "ruleId": "111-bbbbb-ccc",
          "value": true,
        },
        Object {
          "key": "empanada.nin-export-issues-dashboard-gadgets_yb8zi",
          "reason": "RULE_MATCH",
          "ruleId": "111-bcbcbcb-dddd",
          "value": false,
        },
        Object {
          "key": "nav-ux-enable-async-edit-filter-modal",
          "value": true,
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "project.config.pce.project-archival",
          "value": true,
        },
        Object {
          "key": "jsw.frontend.growth.experiments.persistent.upgrade",
          "reason": "FALLTHROUGH",
          "value": "not-enrolled",
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
]
`;

exports[`compressor batchSize argument should compress to one batch if batchSize is larger than events list length 1`] = `
Array [
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "pin-2.0.refinementbar-dropdown",
          "reason": "SIMPLE_EVAL",
          "value": false,
        },
        Object {
          "key": "jsd.queues.keyboard-shortcuts",
          "reason": "ERROR",
          "ruleId": "111-bbbbb-ccc",
          "value": true,
        },
        Object {
          "key": "empanada.nin-export-issues-dashboard-gadgets_yb8zi",
          "reason": "RULE_MATCH",
          "ruleId": "111-bcbcbcb-dddd",
          "value": false,
        },
        Object {
          "key": "nav-ux-enable-async-edit-filter-modal",
          "value": true,
        },
        Object {
          "key": "project.config.pce.project-archival",
          "value": true,
        },
        Object {
          "key": "jsw.frontend.growth.experiments.persistent.upgrade",
          "reason": "FALLTHROUGH",
          "value": "not-enrolled",
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
]
`;

exports[`compressor events argument should compress all normal events 1`] = `
Array [
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "pin-2.0.refinementbar-dropdown",
          "reason": "SIMPLE_EVAL",
          "value": false,
        },
        Object {
          "key": "jsd.queues.keyboard-shortcuts",
          "reason": "ERROR",
          "ruleId": "111-bbbbb-ccc",
          "value": true,
        },
        Object {
          "key": "empanada.nin-export-issues-dashboard-gadgets_yb8zi",
          "reason": "RULE_MATCH",
          "ruleId": "111-bcbcbcb-dddd",
          "value": false,
        },
        Object {
          "key": "nav-ux-enable-async-edit-filter-modal",
          "value": true,
        },
        Object {
          "key": "project.config.pce.project-archival",
          "value": true,
        },
        Object {
          "key": "jsw.frontend.growth.experiments.persistent.upgrade",
          "reason": "FALLTHROUGH",
          "value": "not-enrolled",
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
]
`;

exports[`compressor events argument should default to uncompressed events for invalid events and skip those with no attributes 1`] = `
Array [
  Object {
    "action": "exposed",
    "actionSubject": "features",
    "attributes": Object {
      "evaluations": Array [
        Object {
          "key": "empanada.nin-export-issues-dashboard-gadgets_yb8zi",
          "reason": "RULE_MATCH",
          "ruleId": "111-bcbcbcb-dddd",
          "value": false,
        },
      ],
    },
    "eventType": "operational",
    "source": "@atlassian/exposure-events-compressor",
    "tags": Array [
      "measurement",
      "autoExposure",
    ],
  },
  Object {
    "action": "exposed",
    "actionSubject": "feature",
    "env": "dev",
    "eventType": "operational",
    "origin": "web",
    "platform": "web",
    "product": "jira",
    "sessionId": "1605514166708",
    "source": "@atlaskit/feature-flag-client",
    "subproduct": "platform",
    "tabId": "066ed234-52b1-43ad-8fd3-b2eddb8e2824",
    "tags": Array [],
    "taskSessions": Object {},
    "tenantId": "e433828a-6e8c-419f-bfb4-dac258ea2886",
    "tenantIdType": "cloudId",
    "userIdType": "atlassianAccount",
    "version": "1.0.0",
  },
  Object {
    "action": "exposed",
    "actionSubject": "feature",
    "env": "dev",
    "eventType": "operational",
    "origin": "web",
    "platform": "web",
    "product": "jira",
    "sessionId": "1605514166708",
    "source": "@atlaskit/feature-flag-client",
    "subproduct": "platform",
    "tabId": "066ed234-52b1-43ad-8fd3-b2eddb8e2824",
    "tags": Array [
      "autoExposure",
      "measurement",
    ],
    "taskSessions": Object {},
    "tenantId": "e433828a-6e8c-419f-bfb4-dac258ea2886",
    "tenantIdType": "cloudId",
    "userIdType": "atlassianAccount",
    "version": "1.0.0",
  },
  Object {
    "action": "exposed",
    "actionSubject": "feature",
    "attributes": Object {
      "flagKey": "empanada.nin-export-issues-dashboard-gadgets_yb8zi",
      "reason": "RULE_MATCH",
      "ruleId": "111-bcbcbcb-dddd",
      "value": false,
    },
    "env": "dev",
    "eventType": "track",
    "origin": "web",
    "platform": "web",
    "product": "jira",
    "sessionId": "1605514166708",
    "source": "@atlaskit/feature-flag-client",
    "subproduct": "platform",
    "tabId": "066ed234-52b1-43ad-8fd3-b2eddb8e2824",
    "tags": Array [
      "autoExposure",
      "measurement",
    ],
    "taskSessions": Object {},
    "tenantId": "e433828a-6e8c-419f-bfb4-dac258ea2886",
    "tenantIdType": "cloudId",
    "userIdType": "atlassianAccount",
    "version": "1.0.0",
  },
]
`;

exports[`compressor events argument should return empty list if events list is empty 1`] = `Array []`;
